src/main/java/com/amazon/opendistroforelasticsearch/sql/executor/cursor/CursorResultExecutor.java

```diff
@@ -0,0 +1,146 @@
+            if (status > 399 && status < 500) {
+                Metrics.getInstance().getNumericalMetric(MetricName.FAILED_REQ_COUNT_CUS).increment();
+            } else if (status > 499) {
+                Metrics.getInstance().getNumericalMetric(MetricName.FAILED_REQ_COUNT_SYS).increment();
+            }
+            LOG.error("Error completing cursor request", e);
+            channel.sendResponse(new BytesRestResponse(channel, e));
+        }
+    }
+
+    public String execute(Client client, Map<String, String> params) throws Exception {
+        /**
+         * All cursor's are of the form <cursorType>:<base64 encoded cursor>
+         * The serialized form before encoding is upto Cursor implementation
+         */
+        String[] splittedCursor = cursorId.split(":");
+
+        if (splittedCursor.length!=2) {
+            throw new VerificationException("Not able to parse invalid cursor");
+        }
+
+        String type = splittedCursor[0];
+        CursorType cursorType = CursorType.getById(type);
+
+        if (cursorType!=CursorType.NULL) {
+            switch(cursorType) {
+                case DEFAULT:
+                    DefaultCursor defaultCursor = DefaultCursor.from(splittedCursor[1]);
+                    return handleDefaultCursorRequest(client, defaultCursor);
+                case AGGREGATION:
+                case JOIN:
+                default: throw new VerificationException("Unsupported cursor");
+            }
+        }
+
+        throw new VerificationException("Not able to parse invalid cursor");
+    }
+
+    private String handleDefaultCursorRequest(Client client, DefaultCursor cursor) {
+        String previousScrollId = cursor.getScrollId();
+        LocalClusterState clusterState = LocalClusterState.state();
+        TimeValue scrollTimeout = clusterState.getSettingValue(CURSOR_KEEPALIVE);
+        SearchResponse scrollResponse = client.prepareSearchScroll(previousScrollId).setScroll(scrollTimeout).get();
+        SearchHits searchHits = scrollResponse.getHits();
+        SearchHit[] searchHitArray = searchHits.getHits();
+        String newScrollId = scrollResponse.getScrollId();
+
+        int rowsLeft = (int) cursor.getRowsLeft();
+        int fetch = cursor.getFetchSize();
+
+        if (rowsLeft < fetch && rowsLeft < searchHitArray.length) {
+            /**
+             * This condition implies we are on the last page, and we might need to truncate the result from SearchHit[]
+             * Avoid truncating in following two scenarios
+             * 1. number of rows to be sent equals fetchSize
+             * 2. size of SearchHit[] is already less that rows that needs to be sent
+             *
+             * Else truncate to desired number of rows
+             */
+            SearchHit[] newSearchHits = Arrays.copyOf(searchHitArray, rowsLeft);
+            searchHits = new SearchHits(newSearchHits, searchHits.getTotalHits(), searchHits.getMaxScore());
+        }
```

Comment:
Good suggestion, will do, wasn't apparent initially.

Reply:
I realized that though we can skip unwanted `DataRows.Row` using `limit`, but this will apply to final flattened row, and will impact correctness result for nested queries.

For example:
If the final page has `rowLeft` as 2, implying two un-nested doc, and assuming each doc has atleast 3 nested docs. If use above logic there will be 2 `Row`, instead of 8 that we expect.